### ISSUE TITLE
How to pass more than 1 pillar variable ?

### ISSUE THREAD

**Issue by amalaguti** [jenkinsci/saltstack-plugin#15]
Do you know how to pass more than a single pillar value when using a runner (orchestration)?

salt-run state.orchestrate orch.orc_ddns-v2 pillar='{IPallocation: 1.2.3.4, hostname: mynewhost}'

Client interface: runner
mods: orch.orc_ddns-v2
Function: state.orchestrate

I tried multiple combinations in Arguments and KWArguments but nothing worked.
Using the Set Pillar Data looks like it only takes 1 entry.

**Comment by mchugh19** [comment]
Unfortunately that is an area that will require more work. Because of how the runner formats pillar data into the request, the args and kwargs are unused; hence the single entry for the pillar fields.

This is currently at the top of the to-do list. So it's now just a matter of figuring out the Jenkins way of handling repeatable pillar data.

**Comment by amalaguti** [comment]
Thanks Christian for the quick response ! It would be great ! I keep watching for updates. thank you!

**Comment by amalaguti** [comment]
Chris,

I modified my pillar to have a single one object with ip and hostname attributes.
pillar='{DDNSallocation: [ip: 10.10.10.10, hostname: newhost]}’

{% set hostname = salt['pillar.get']('DDNSallocation:hostname','newhost') %}
{% set ip = salt['pillar.get']('DDNSallocation:ip','1.1.1.1') %}

How should I reference it in the Set Pillar Data key field ?
Tried with a few combinations with no luck.
DDNSallocation:ip
DDNSallocation.ip
'DDNSallocation:ip'
‘DDNSallocation.ip'

**Comment by mchugh19** [comment]
Previously json data was being interpreted as a string. I've just worked up a beta to handle pillar values as json objects. 

Please test [this](https://github.com/jenkinsci/saltstack-plugin/releases/tag/1.7.0-beta6)

You should be able to check "Set Pillar Data" with a key of DDNSallocation and a value of ```{"ip":"10.10.10.10", "hostname":"newhost"}```. Note that I changed your [] array to {}. This is a quick proof of concept and I'll fix that bit next.

**Comment by amalaguti** [comment]
Failed… but at least it throws a message.
Tried both with {} and []
{"ip":"10.10.10.10", "hostname":”VMnew”}

["ip":"10.10.10.10", "hostname":"VMnew”]

Now in addition it does fail with a single pillar, same message. It was working in the previous version with a single pillar.


Building in workspace /var/lib/jenkins/jobs/Salt DDNS Orchestration/workspace
ERROR: Build step failed with exception
net.sf.json.JSONException <http://stacktrace.jenkins-ci.org/search?query=net.sf.json.JSONException>: JSONObject["data"] is not a JSONObject.
	at net.sf.json.JSONObject.getJSONObject(JSONObject.java:2006) <http://stacktrace.jenkins-ci.org/search/?query=net.sf.json.JSONObject.getJSONObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:174) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateInnerJsonObject(Utils.java:211) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateInnerJsonObject&entity=method>
	at com.waytta.Utils.validateFunctionCall(Utils.java:159) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.Utils.validateFunctionCall&entity=method>
	at com.waytta.SaltAPIBuilder.perform(SaltAPIBuilder.java:375) <http://stacktrace.jenkins-ci.org/search/?query=com.waytta.SaltAPIBuilder.perform&entity=method>
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20) <http://stacktrace.jenkins-ci.org/search/?query=hudson.tasks.BuildStepMonitor$1.perform&entity=method>
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:782) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.AbstractBuild$AbstractBuildExecution.perform&entity=method>
	at hudson.model.Build$BuildExecution.build(Build.java:205) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.Build$BuildExecution.build&entity=method>
	at hudson.model.Build$BuildExecution.doRun(Build.java:162) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.Build$BuildExecution.doRun&entity=method>
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:534) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.AbstractBuild$AbstractBuildExecution.run&entity=method>
	at hudson.model.Run.execute(Run.java:1738) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.Run.execute&entity=method>
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.FreeStyleBuild.run&entity=method>
	at hudson.model.ResourceController.execute(ResourceController.java:98) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.ResourceController.execute&entity=method>
	at hudson.model.Executor.run(Executor.java:410) <http://stacktrace.jenkins-ci.org/search/?query=hudson.model.Executor.run&entity=method>
Build step 'Send a message to Salt API' marked build as failure
Finished: FAILURE

**Comment by mchugh19** [comment]
Interesting! Could you please send me the json output from the saltapi?

[Here](https://gist.github.com/mchugh19/b7a5084c3a5395bf03ac) are a couple of curl commands that should log in to the saltapi, and then kick off a salt-run job. Please replace the localhost, USERNAME,  PASSWORD, and orchestration.refresh-apache parts with the appropriate values for your environment. 

Thanks!

**Comment by amalaguti** [comment]
Here you have it Chris. 
Looks like it executed fine, the orchestration made the call to the states and from the states a couple of scripts are called, all with proper pillar info set, and record was created in DNS.

**Comment by mchugh19** [comment]
@amalaguti Can you send the json again? Looks like it was missing off that last reply. 

Thanks much!

**Comment by amalaguti** [comment]
Just updated the previous comment uploading a .zip file.

**Comment by mchugh19** [comment]
Woo! Obviously your case triggered a bug. A test has been added and the code improved. Thanks!

[This](https://github.com/jenkinsci/saltstack-plugin/releases/tag/1.7.0-beta7) beta release should fully implement your request for json as a pillar value in a salt-run job. 

Both
```{"ip":"10.10.10.10", "hostname":"VMnew"}```
and
```[{"ip":"10.10.10.10", "hostname":"VMnew"}]```
should now be properly handled. (```["ip":"10.10.10.10", "hostname":"VMnew"]``` is invalid json)

**Comment by amalaguti** [comment]
Chris,
sorry, not sure how I should set the pillar info in the Jenkins build page.

Should I use the Set Pillar Data fields or the Arguments field.
My original intention was to use the Arguments field: pillar={"ip":"10.10.10.10", "hostname":"VMnew”}

is this the way you expect it to work ?

**Comment by mchugh19** [comment]
No problem. Salt runner handles pillar data differently from the standard local client, hence the extra pillar data attributes. I attempted to document the difference on [the plugin wiki](https://wiki.jenkins-ci.org/display/JENKINS/saltstack-plugin)

For your cli command of
```
salt-run state.orchestrate orch.orc_ddns-v2 pillar='{"DDNSallocation":{"IPallocation": "1.2.3.4", "hostname": "mynewhost"}}'
```

Would translate to:
```
Client Interface: runner
mods: orch.orc_ddns-v2
Set Pillar Data: checked
Pillar key: DDNSallocation
Pillar value: {"IPallocation": "1.2.3.4", "hostname": "mynewhost"}
Target: notused
Target Type: list (also unused so this doesn't matter)
Function: state.orchestrate
```

**Comment by mchugh19** [comment]
Changes released in 1.7.0
